In [104]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib as plt
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import classification_report

### 2. Given the following confusion matrix, evaluate (by hand) the model's performance.


|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |




|               |            | 
|:------------  |-----------:|
|Precesion      |       .867 |
|Recall         |       .779 |
|Accuracy       |       .8   |



- In the context of this problem, what is a false positive?
        
        Cat is dog

- In the context of this problem, what is a false negative?

        Dog is cat
        
- How would you describe this model?

        

### 3. You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.Use the predictions dataset and pandas to help answer the following questions:

- An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?




In [3]:
ducks= pd.read_csv('c3.csv')

In [4]:
ducks

actual     model1     model2     model3
0    No Defect  No Defect     Defect  No Defect
1    No Defect  No Defect     Defect     Defect
2    No Defect  No Defect     Defect  No Defect
3    No Defect     Defect     Defect     Defect
4    No Defect  No Defect     Defect  No Defect
..         ...        ...        ...        ...
195  No Defect  No Defect     Defect     Defect
196     Defect     Defect  No Defect  No Defect
197  No Defect  No Defect  No Defect  No Defect
198  No Defect  No Defect     Defect     Defect
199  No Defect  No Defect  No Defect     Defect

[200 rows x 4 columns]

In [11]:
# Lets start by crosstabing each model into its own table with actual
model_1 = pd.crosstab(ducks.actual,ducks.model1)
model_2 = pd.crosstab(ducks.actual,ducks.model2)
model_3 = pd.crosstab(ducks.actual,ducks.model3)

In [7]:
model_1

model1     Defect  No Defect
actual                      
Defect          8          8
No Defect       2        182

In [9]:
model_2

model2     Defect  No Defect
actual                      
Defect          9          7
No Defect      81        103

In [10]:
model_3

model3     Defect  No Defect
actual                      
Defect         13          3
No Defect      86         98

 The internal team wants to investigate the cause of the manufacturing defects. They say that they want to identify as many of the ducks that have a defect as possible. The metric we should use here is Recall. It compares true positive to the false negative. Our true positive in this case would be that the Duck is a defect our False negative would be that our duck is classified as having no defect when it is defective. We want our recall to be as accurate as possible to minimize the amount of defective units that make it through production

In [17]:
subset =  ducks[ducks.actual == 'Defect']
model_recall_1 = (subset.model1 == subset.actual).mean()

In [19]:
model_recall_1


0.5

In [20]:
model_recall_2 = (subset.model2 == subset.actual).mean()

In [21]:
model_recall_2

0.5625

In [22]:
model_recall_3 = (subset.model3 == subset.actual).mean()

In [23]:
model_recall_3

0.8125

Model 3 has the highest accuracy for recall so we should be using it to catch defects in production


#### Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

An argument could be made to use precesion here but ultimatley i think accuracy is the way to go especially with a 95% accurate model.

In [28]:
model1_accuracy = (ducks.model1 == ducks.actual).mean()
model1_accuracy 

0.95

In [26]:
model2_accuracy = (ducks.model2 == ducks.actual).mean()
model2_accuracy

0.56

In [27]:
model3_accuracy = (ducks.model3 == ducks.actual).mean()
model3_accuracy 

0.555

### You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).Several models have already been developed with the data, and you can find their results here.Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

- In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?


In [32]:
df = pd.read_csv('gives_you_paws.csv')

In [36]:
df.actual.value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

In [37]:
# Creat a baseline always populating dog
df['baseline']= 'dog'

In [43]:
# create our models
model_1 = pd.crosstab(df.actual,df.model1)
model_2 = pd.crosstab(df.actual,df.model2)
model_3 = pd.crosstab(df.actual,df.model3)

In [44]:
model_1

model1   cat   dog
actual            
cat     1423   323
dog      640  2614

In [47]:
model1_accuracy = (df.model1 == df.actual).mean()
model1_accuracy

0.8074

In [48]:
model2_accuracy = (df.model2 == df.actual).mean()
model2_accuracy

0.6304

In [49]:
model3_accuracy = (df.model3 == df.actual).mean()
model3_accuracy

0.5096

In [50]:
baseline_accuracy = (df.baseline == df.actual).mean()
baseline_accuracy

0.6508

|               |  Accuracy  | 
|:------------  |-----------:|
|baseline       |     0.6508 |
|model 1        |     0.8074 |
|model 2        |     0.6304 |
|model 3        |     0.5096 |


### Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?

In [53]:
# Lets find the best precision model
subset = df[df.model1 == 'dog']
model1_precision = (subset.model1 == subset.actual).mean()

In [54]:
model1_precision

0.8900238338440586

In [56]:
subset = df[df.model2 == 'dog']
model2_precision = (subset.model2 == subset.actual).mean()

In [57]:
model2_precision

0.8931767337807607

In [58]:
subset = df[df.model3 == 'dog']
model3_precision = (subset.model3 == subset.actual).mean()

In [59]:
model3_precision

0.6598883572567783

I would use model 1 accuracy for phase 1 and model 2 for precision during phase 2

### Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?

In [116]:
subset = df[df.actual == 'cat']
model1_recall = (subset.model1 == subset.actual).mean()

In [117]:
model1_recall

0.8150057273768614

In [118]:

model2_recall = (subset.model2 == subset.actual).mean()

In [120]:
model2_recall

0.8906071019473081

In [121]:
model3_recall = (subset.model3 == subset.actual).mean()

In [122]:
model3_recall

0.5114547537227949

model 1 for accuracy and for precision 

#### Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

sklearn.metrics.accuracy_score

sklearn.metrics.precision_score

sklearn.metrics.recall_score

sklearn.metrics.classification_report

In [69]:
accuracy_score(df.actual, df.model1)

0.8074

In [70]:
accuracy_score(df.actual, df.model2)

0.6304

In [71]:
accuracy_score(df.actual, df.model3)

0.5096

In [72]:
accuracy_score(df.actual, df.baseline)

0.6508

|               |  Accuracy  | 
|:------------  |-----------:|
|baseline       |     0.6508 |
|model 1        |     0.8074 |
|model 2        |     0.6304 |
|model 3        |     0.5096 |

In [87]:
 precision_score(df.actual, df.model1, pos_label = 'dog')

0.8900238338440586

In [88]:
 precision_score(df.actual, df.model2, pos_label = 'dog')

0.8931767337807607

In [90]:
 precision_score(df.actual, df.model3, pos_label = 'dog')

0.6598883572567783

In [91]:
 precision_score(df.actual, df.baseline, pos_label = 'dog')

0.6508

|               | precision  | 
|:------------  |-----------:|
|baseline       |   0.6508   |
|model 1        |   0.8900   |
|model 2        |    0.8931  |
|model 3        |     0.6598 |

In [98]:
recall_score(df.actual, df.model1, pos_label = 'dog')

0.803318992009834

In [99]:
recall_score(df.actual, df.model2, pos_label = 'dog')

0.49078057775046097

In [100]:
recall_score(df.actual, df.model3, pos_label = 'dog')

0.5086047940995697

In [102]:
 recall_score(df.actual, df.baseline, pos_label = 'dog')

1.0

In [109]:
target_names = ['cat', 'dog']
print(classification_report(df.actual, df.model1, target_names=target_names))

              precision    recall  f1-score   support

         cat       0.69      0.82      0.75      1746
         dog       0.89      0.80      0.84      3254

    accuracy                           0.81      5000
   macro avg       0.79      0.81      0.80      5000
weighted avg       0.82      0.81      0.81      5000



In [110]:
print(classification_report(df.actual, df.model2, target_names=target_names))

              precision    recall  f1-score   support

         cat       0.48      0.89      0.63      1746
         dog       0.89      0.49      0.63      3254

    accuracy                           0.63      5000
   macro avg       0.69      0.69      0.63      5000
weighted avg       0.75      0.63      0.63      5000



In [111]:
print(classification_report(df.actual, df.model3, target_names=target_names))

              precision    recall  f1-score   support

         cat       0.36      0.51      0.42      1746
         dog       0.66      0.51      0.57      3254

    accuracy                           0.51      5000
   macro avg       0.51      0.51      0.50      5000
weighted avg       0.55      0.51      0.52      5000



In [112]:
print(classification_report(df.actual, df.baseline, target_names=target_names))

              precision    recall  f1-score   support

         cat       0.00      0.00      0.00      1746
         dog       0.65      1.00      0.79      3254

    accuracy                           0.65      5000
   macro avg       0.33      0.50      0.39      5000
weighted avg       0.42      0.65      0.51      5000



/opt/homebrew/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/homebrew/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/homebrew/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
